src/mpi/comm/commutil.c

```diff
@@ -478,6 +503,49 @@ int MPIR_Comm_commit(MPIR_Comm * comm)
         comm->hierarchy_kind = MPIR_COMM_HIERARCHY_KIND__PARENT;
     }
 
+    /* Topology information collection for selection logic */
+    mpi_errno = MPID_Get_max_node_id(comm, &max_node_id);
+    if (mpi_errno)
+        MPIR_ERR_POP(mpi_errno);
+    MPIR_Assert(max_node_id >= 0);
+    nodemap = MPL_calloc((max_node_id + 1), sizeof(int), MPL_MEM_OTHER);
+    for (i = 0; i < comm->remote_size; ++i) {
+        mpi_errno = MPID_Get_node_id(comm, i, &node_id);
+        if (mpi_errno)
+            MPIR_ERR_POP(mpi_errno);
+        MPIR_Assert(node_id >= 0);
+        nodemap[node_id]++;
+        if (comm->max_num_procs < nodemap[node_id]) {
+            comm->max_num_procs = nodemap[node_id];
+        }
+    }
+
+    for (i = 0; i <= max_node_id; i++) {
+        if (nodemap[i] < comm->max_num_procs) {
+            comm->uniform_ranks = 0;
+            break;
+        }
+    }
+
+    if (strcmp(MPIR_CVAR_ALLGATHER_INTRA_AUTO_TUNE, "")) {
+        fp = fopen(MPIR_CVAR_ALLGATHER_INTRA_AUTO_TUNE, "r");
+        if (fp) {
```

Comment:
Do we really want to open/read/parse the file for every new communicator?

Reply:
@hzhou Any suggestions?